flink-streaming-java/src/test/java/org/apache/flink/streaming/api/graph/AdaptiveGraphManagerTest.java

```diff
@@ -0,0 +1,354 @@
+
+        DataStream<Tuple2<String, String>> result =
+                input.keyBy(x -> x.f0)
+                        .map(
+                                new MapFunction<Tuple2<String, String>, Tuple2<String, String>>() {
+
+                                    @Override
+                                    public Tuple2<String, String> map(
+                                            Tuple2<String, String> value) {
+                                        return value;
+                                    }
+                                });
+
+        result.sinkTo(new DiscardingSink<>());
+        StreamGraph streamGraph = env.getStreamGraph();
+        JobGraph jobGraph1 = generateJobGraphInLazilyMode(streamGraph);
+        JobGraph jobGraph2 = StreamingJobGraphGenerator.createJobGraph(streamGraph);
+        assertThat(isJobGraphEquivalent(jobGraph1, jobGraph2)).isEqualTo(true);
+    }
+
+    @Test
+    void testSourceChain() {
+        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+        env.setMaxParallelism(100);
+        env.setParallelism(100);
+        MultipleInputTransformation<Long> transform =
+                new MultipleInputTransformation<>(
+                        "mit", new UnusedOperatorFactory(), Types.LONG, -1);
+
+        Transformation<Long> input1 =
+                env.fromSource(
+                                new NumberSequenceSource(1, 2),
+                                WatermarkStrategy.noWatermarks(),
+                                "input1")
+                        .setParallelism(100)
+                        .getTransformation();
+        Transformation<Long> input2 =
+                env.fromSource(
+                                new NumberSequenceSource(1, 2),
+                                WatermarkStrategy.noWatermarks(),
+                                "input2")
+                        .setParallelism(1)
+                        .getTransformation();
+        Transformation<Long> input3 =
+                env.fromSource(
+                                new NumberSequenceSource(1, 2),
+                                WatermarkStrategy.noWatermarks(),
+                                "input3")
+                        .setParallelism(1)
+                        .getTransformation();
+        transform.addInput(input1);
+        transform.addInput(input2);
+        transform.addInput(input3);
+        transform.setChainingStrategy(ChainingStrategy.HEAD_WITH_SOURCES);
+        DataStream<Long> dataStream = new DataStream<>(env, transform);
+        // do not chain with sink operator.
+        dataStream.rebalance().sinkTo(new DiscardingSink<>()).name("sink");
+        env.addOperator(transform);
+        StreamGraph streamGraph = env.getStreamGraph();
+        streamGraph.setDynamic(true);
+        JobGraph jobGraph = createJobGraph(streamGraph);
+        assertThat(jobGraph.getVerticesSortedTopologicallyFromSources().size()).isEqualTo(4);
```

Comment:
Only one source can be chained in this case? input3 will not be included in the multi-input chain?

Reply:
> Only one source can be chained in this case? input3 will not be included in the multi-input chain?

Yes, the parallelism of input2 and input3 is 1, and only the parallelism of input1 meets the conditions of sourceChain.